### PR TITLE
Fix istft fake tensor length padding

### LINKED ISSUE #183886: torch.compile fails on torch.istft with RuntimeError when making fake tensor call
### 🐛 Describe the bug

`torch.compile` with the Inductor backend fails on a module using `torch.istft`.

Eager execution succeeds and returns an output with shape `(3, 352)`, while the compiled version fails during fake tensor execution with: `TorchRuntimeError RuntimeError when making fake tensor call`

```
import warnings
import torch
import torch.nn as nn

class M(nn.Module):
    def __init__(self):
        super().__init__()
        self.n_fft = 64
        self.hop_length = 32
        self.length = 352
        self.register_buffer("window", torch.hann_window(self.n_fft), persistent=False)

    def forward(self, spec):
        return torch.istft(
            spec,
            n_fft=self.n_fft,
            hop_length=self.hop_length,
            window=self.window,
            length=self.length,
        )

model = M().eval()
spec = torch.view_as_complex(torch.randn(3, 33, 10, 2))

try:
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        eager_out = model(spec)
    eager = f"OK {tuple(eager_out.shape)}"
except Exception as exc:
    eager = f"{type(exc).__name__} {str(exc).splitlines()[0]}"

try:
    compiled_out = torch.compile(model, backend="inductor")(spec)
    compiled = f"OK {tuple(compiled_out.shape)}"
except Exception as exc:
    compiled = f"{type(exc).__name__} {str(exc).splitlines()[0]}"

print("spec_shape:", tuple(spec.shape))
print("length:", model.length)
print("eager:", eager)
print("compiled:", compiled)

if eager.startswith("OK") and not compiled.startswith("OK"):
    raise SystemExit(0)

raise SystemExit(1)
```

### Error logs

```
spec_shape: (3, 33, 10)
length: 352
eager: OK (3, 352)
compiled: TorchRuntimeError RuntimeError when making fake tensor call
```

### Versions

PyTorch version: 2.13.0.dev20260513+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: 18.1.3 (1ubuntu1)
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.11.15 (main, Mar 11 2026, 17:20:07) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.17.0-20-generic-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration:
GPU 0: NVIDIA RTX 6000 Ada Generation
GPU 1: NVIDIA RTX 6000 Ada Generation

Nvidia driver version: 570.211.01
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] torch==2.13.0.dev20260513+cpu
[conda] numpy                      2.4.4                   pypi_0           pypi
[conda] torch                      2.13.0.dev20260513+cpu  pypi_0           pypi

cc @mruberry @chauhang @penguinwu @eellison @aorenste @bdhirsh @bobrenjc93

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #184532

Use slice semantics in the Python istft ref so fake tensor propagation matches the native implementation when a requested length extends past the overlap-add buffer before zero padding. Add a Dynamo regression test for the torch.compile failure.

Fixes #183886

Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98 @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by oulgen — APPROVED** [review]
